src/dune_rules/module_compilation.ml

```diff
@@ -86,26 +97,35 @@ let build_cm cctx ~precompiled_cmi ~cm_kind (m : Module.t)
         (* If there is no mli, [ocamlY -c file.ml] produces both the .cmY and
            .cmi. We choose to use ocamlc to produce the cmi and to produce the
            cmx we have to wait to avoid race conditions. *)
-        | Cmo, None, false ->
-          let+ () = copy_interface ~dir ~obj_dir ~sctx m in
-          ([], [], [ Obj_dir.Module.cm_file_exn obj_dir m ~kind:Cmi ])
-        | Cmo, None, true | (Cmo | Cmx), _, _ ->
+        | (Ocaml Cmo | Melange Cmj), None, false ->
+          let+ () = copy_interface ~dir ~obj_dir ~sctx ~cm_kind m in
+          let cmi_kind = Lib_mode.Cm_kind.cmi cm_kind in
+          ([], [], [ Obj_dir.Module.cm_file_exn obj_dir m ~kind:cmi_kind ])
+        | (Ocaml Cmo | Melange Cmj), None, true
+        | (Ocaml (Cmo | Cmx) | Melange Cmj), _, _ ->
+          let cmi_kind = Lib_mode.Cm_kind.cmi cm_kind in
           Memo.return
-            ( force_read_cmi src
-            , [ Path.build (Obj_dir.Module.cm_file_exn obj_dir m ~kind:Cmi) ]
+            ( force_read_cmi ~cm_kind src
+            , [ Path.build (Obj_dir.Module.cm_file_exn obj_dir m ~kind:cmi_kind)
+              ]
             , [] )
-        | Cmi, _, _ ->
-          let+ () = copy_interface ~dir ~obj_dir ~sctx m in
+        | (Ocaml Cmi | Melange Cmi), _, _ ->
+          let+ () = copy_interface ~dir ~obj_dir ~sctx ~cm_kind m in
           ([], [], []))
   in
   let other_targets =
     match cm_kind with
-    | Cmx -> (
+    | Ocaml Cmx -> (
       match phase with
       | Some Compile -> linear :: other_targets
       | Some Emit -> other_targets
       | Some All | None -> obj :: other_targets)
-    | Cmi | Cmo -> other_targets
+    | Ocaml (Cmi | Cmo) | Melange Cmi -> other_targets
+    | Melange Cmj ->
+      let melange_js =
+        Obj_dir.Module.obj_file obj_dir m ~kind:(Melange Cmj) ~ext:".js"
+      in
+      melange_js :: other_targets
```

Comment:
I forgot to ask this previously, but can't we setup the js and cmj targets separately?

Reply:
It is possible, but the results are a bit messy as the `Js` artifact kind needs to start being tracked inside Dune.

As I did not think that addition should block this PR, I created a new one to discuss the changes required to do this, please take a look: https://github.com/jchavarri/dune/pull/6